module/data/fields/advantage-mode-field.mjs

```diff
@@ -14,46 +14,115 @@ export default class AdvantageModeField extends foundry.data.fields.NumberField
   /* -------------------------------------------- */
 
   /**
-   * Number of advantage modifications.
-   * @type {number}
+   * Allowed advantage mode values.
+   * @type {number[]}
    */
-  #advantage;
+  static #values = [-1, 0, 1];
 
   /* -------------------------------------------- */
 
   /**
-   * Number of disadvantage modifications.
-   * @type {number}
+   * Advantage sources.
+   * @type {Map<string, Map<string, number>>}
    */
-  #disadvantage;
+  static #advantages = new Map();
 
   /* -------------------------------------------- */
 
-  /** @inheritDoc */
-  initialize(value, model, options={}) {
-    this.#advantage = Number(value === 1);
-    this.#disadvantage = Number(value === -1);
-    return value;
+  /**
+   * Disadvantage sources.
+   * @type {Map<string, Map<string, number>>}
+   */
+  static #disadvantages = new Map();
+
+  /* -------------------------------------------- */
+
+  /**
+   * Add one to the count of advantages or disadvantages.
+   * @param {Actor5e} actor           The actor gaining a source of advantage or disadvantage.
+   * @param {string} path             The path to the property affected.
+   * @param {boolean} [isAdvantage]   Is this an instance of advantage or disadvantage?
+   */
+  static addSource(actor, path, isAdvantage=true) {
```

Comment:
The static model schemas do make accomplishing this a bit more difficult. We do have an instance available to us though, if we don't want to do everything in static scope with maps like this. Active Effects are already a part of data preparation, so I don't think there's any conceptual issue with us applying some additional prepared data that isn't the exact property that was targeted.

For example, where we wanted an instance-level `advantage` property, we could store it on the model instance itself, as a sibling to the targeted property.

```js
const roll = foundry.utils.getProperty(model, change.key.substring(0, change.key.lastIndexOf(".")));
roll.advantage++;
```

The property will be removed/reset when the Actor is re-prepared, like all derived properties. This does rely somewhat on a predictable data structure though.

We could also put these maps on the model instead, and track them via `fieldPath` in a similar way to this PR. Something like `actor.system.advantages`, similar to the original proposal in #809.

I think I prefer keeping the data closer to the targeted field though, which allows the PR to function mostly as it did before.

I did take a look at your original commit in this PR, and the logic looks good to me. So we would want to integrate whichever of the above proposals we settle on with your original logic. Sorry that I was not clearer in my comments on your other PR.

Also, apologies if this is feeling like too much whiplash. I think I have enough that I could make any final changes myself now if you'd prefer. Any thoughts you have on the above are welcome though.

Reply:
I'll force a commit and we'll see what comes of it.